### PR TITLE
Linter_Bears_Advanced.rst: Explains return `None`

### PR DESCRIPTION
**Pull request by Vedarth**
Mentions how returning `None` inside `generate_config` works

Closes https://github.com/coala/coala/issues/4119

**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [x] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [x] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `cobot mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

### CODE REVIEW

**Comment by Vedarth** [comment]
@jschwarzwalder kindly review this pr.

**Comment by jschwarzwalder** [comment]
Thank you for starting the proposed change @Vedarth. 

I like that you are encouraging developers to return None, but I'm not sure if it is clear how to use create_arguments() and generate_config() together when they need to make a config file. 

When I first read the change I had the question: Does the developer call create_arguments() or generate_config() with their own parameters, or do they write the code themselves? If they are only calling these, Can the developer change the return value? It seems like this article is directing developers to override existing functions. Is this correct?


I was also mildly confused on which function the note is referring. 
Above this addition referring to create_arguments() it states that _if you return `None` then no configuration file is generated._ 
Later it states _By default, no configuration file is generated._, but I"m not sure the subject this latter statement is referring. 

> If create_arguments(filename, file, config_file) returns a config file, then, returning a config file in def generate_config(filename, file) is optional.

It sounds like you are trying to say that you can return a config file in create_arguments() or in generate_config().

> Returning None will avoid generating a useless config file, which will increase the performance substantially. 

Is this referring to create_arguments() or generate_config()?

Can you revise the paragraph to be clearer for developers on when and where they should return None. You've done a solid job of explaining why.  

Thank you for your contribution. 😸

**Comment by Makman2** [comment]
unack e38d0e6

**Comment by Vedarth** [comment]
@Makman2 I could use some more clarification regarding this.
And @jschwarzwalder Thanks for your helpful and detailed review. Your words are very encouraging. I will keep your feedback in mind in my future commits, hopefully only one of those will be related to this issue 😛.

**Comment by Makman2** [comment]
@Vedarth See https://github.com/coala/coala/issues/4119#issuecomment-347625520 ;)

**Comment by Vedarth** [comment]
Sorry for being so late! I was busy with my university exams. I will try to do the required changes asap.

**Comment by Vedarth** [comment]
@jschwarzwalder @Makman2 I am extremely sorry for updating the pr after so long. I was busy with university related stuff. I assure you I will implement your next feedback way more swiftly. Thanks for keeping faith in me.

File: docs/Developers/Linter_Bears_Advanced.rst [hunk 1]
```diff
@@ -41,6 +41,31 @@ The path of the temporary configuration file can be accessed inside
         def create_arguments(filename, file, config_file):
             return "--use-config", config_file
 
+The ``generate_config()`` should not return ``config_file`` if the user wants
+to use his own tool-specific config, to increase performance. If a config
+file is supplied by user that means he won't need the setting provided by
+coala to generate the config file. Therefore, in such scenarios, it is advised
+to return ``None`` in ``generate_config()``.
+For more information on coala settings:-
+http://api.coala.io/en/latest/Developers/coala_settings.html
```
Comment (Makman2) [inline docs/Developers/Linter_Bears_Advanced.rst]: That reference doesn't really belong here^^

File: docs/Developers/Linter_Bears_Advanced.rst [hunk 2]
```diff
@@ -41,6 +41,31 @@ The path of the temporary configuration file can be accessed inside
         def create_arguments(filename, file, config_file):
             return "--use-config", config_file
 
+The ``generate_config()`` should not return ``config_file`` if the user wants
```
Comment (Makman2) [inline docs/Developers/Linter_Bears_Advanced.rst]: Maybe you should move the part

> If you return ``None``, no configuration file is generated.

here, and then start with "A common case where to explicitly return `None` is ..."

**Comment by Makman2** [comment]
unack f79fa6444067f322fe1f7605677800860d838331

File: docs/Developers/Linter_Bears_Advanced.rst (hunk 1 above)
Comment (Vedarth) [reply docs/Developers/Linter_Bears_Advanced.rst]: I thought it might be helpful if the reader doesn't know about the settings in Coala. But I will remove it if you say so 😄

File: docs/Developers/Linter_Bears_Advanced.rst (hunk 2 above)
Comment (Vedarth) [reply docs/Developers/Linter_Bears_Advanced.rst]: Got it 😄

File: docs/Developers/Linter_Bears_Advanced.rst (hunk 1 above)
Comment (Makman2) [reply docs/Developers/Linter_Bears_Advanced.rst]: The reader should, as it's developer's documentation ;)

PS:
coala is always written with a lowercase c ;)

File: docs/Developers/Linter_Bears_Advanced.rst [hunk 3]
```diff
@@ -41,6 +40,30 @@ The path of the temporary configuration file can be accessed inside
...
+tool-specific config, to increase performance. If a config file is supplied
+by user that means he won't need the setting provided by coala to generate
+the config file. Therefore, in such scenarios, it is advised to return
+``None`` in ``generate_config()``.
+
+::
+
+    @linter(executable='...')
+    class MyBear:
+    def generate_config(filename, file, settingA):
```
Comment (Makman2) [inline docs/Developers/Linter_Bears_Advanced.rst]: wrong indentation ;)

File: docs/Developers/Linter_Bears_Advanced.rst [hunk 4]
```diff
@@ -41,6 +40,30 @@ The path of the temporary configuration file can be accessed inside
...
+            config_file = ("value1 = 1\n"
+                           "value2 = 2")
+            return config_file
+        else:
+            return None
+
+    def create_arguments(filename, file, config_file, settingA):
+        if config_file:
+            return "--use-config", config file
+        return args
```
Comment (Makman2) [inline docs/Developers/Linter_Bears_Advanced.rst]: which `args`? You should somehow use `settingA` (and it should be optional :+1:)

File: docs/Developers/Linter_Bears_Advanced.rst [hunk 5]
```diff
@@ -41,6 +40,30 @@ The path of the temporary configuration file can be accessed inside
...
+        if settingA:
+            config_file = ("value1 = 1\n"
+                           "value2 = 2")
+            return config_file
+        else:
+            return None
+
+    def create_arguments(filename, file, config_file, settingA):
+        if config_file:
+            return "--use-config", config file
```
Comment (Makman2) [inline docs/Developers/Linter_Bears_Advanced.rst]: please use single-quotes, usually preferred if there are no single-quotes inside ;)

Comment (Makman2) [inline docs/Developers/Linter_Bears_Advanced.rst]: Bad space inside `config file`

File: docs/Developers/Linter_Bears_Advanced.rst (hunk 4 above)
Comment (Makman2) [inline docs/Developers/Linter_Bears_Advanced.rst]: optional: Maybe write up to:
```python
return ('--use-config', config_file) if config_file else ...
```
?

File: docs/Developers/Linter_Bears_Advanced.rst [hunk 6]
```diff
@@ -41,6 +40,30 @@ The path of the temporary configuration file can be accessed inside
...
+
+    @linter(executable='...')
+    class MyBear:
+    def generate_config(filename, file, settingA):
+        if settingA:
+            config_file = ("value1 = 1\n"
+                           "value2 = 2")
+            return config_file
+        else:
+            return None
```
Comment (Makman2) [inline docs/Developers/Linter_Bears_Advanced.rst]: hmm this is not quite right, have you read bears' code where your described feature is used? (E.g. https://github.com/coala/coala-bears/blob/1a7c514aff65d804d86ca4df6ea2b51adc5403cc/bears/css/CSSCombBear.py)

File: docs/Developers/Linter_Bears_Advanced.rst [hunk 7]
```diff
@@ -41,6 +40,30 @@ The path of the temporary configuration file can be accessed inside
         def create_arguments(filename, file, config_file):
             return "--use-config", config_file
 
+If you return ``None``, no configuration file is generated. A common case
+where to explicitly return ``None`` is when the user wants to use his own
+tool-specific config, to increase performance. If a config file is supplied
+by user that means he won't need the setting provided by coala to generate
+the config file. Therefore, in such scenarios, it is advised to return
+``None`` in ``generate_config()``.
```
Comment (Makman2) [inline docs/Developers/Linter_Bears_Advanced.rst]: Hm this still reads a bit complicated, I'm still not convinced... I'll wrap up my head again for this, please fix the other issues already ;)

File: docs/Developers/Linter_Bears_Advanced.rst (hunk 6 above)
Comment (Vedarth) [reply docs/Developers/Linter_Bears_Advanced.rst]: Should I write it for this specific bear? I thought we have to write a general example.

File: docs/Developers/Linter_Bears_Advanced.rst [hunk 8]
```diff
@@ -41,6 +40,28 @@ The path of the temporary configuration file can be accessed inside
         def create_arguments(filename, file, config_file):
             return "--use-config", config_file
 
+If you return ``None``, no configuration file is generated. A common case
+where to explicitly return ``None`` is when the user wants to use his own
```
Comment (Makman2) [inline docs/Developers/Linter_Bears_Advanced.rst]: maybe rephrase to
```
... is when you want to expose a setting for the user to use his own tool-specific config.
```
To clear the ambiguity between the user and the programmer, who wants what :D

File: docs/Developers/Linter_Bears_Advanced.rst [hunk 9]
```diff
@@ -41,6 +40,28 @@ The path of the temporary configuration file can be accessed inside
...
+tool-specific config. If a config file is supplied by user that means we
+don't have to generate a config file. Generating config file consumes
+resources. Avoiding generating config file increases the performance.
```
Comment (Makman2) [inline docs/Developers/Linter_Bears_Advanced.rst]: ~~If a config file ...~~
-->
```
In case the user specifies such a config file, we can avoid generating one again with ``generate_config`` to reduce I/O load.
```

File: docs/Developers/Linter_Bears_Advanced.rst [hunk 10]
```diff
@@ -41,6 +40,28 @@ The path of the temporary configuration file can be accessed inside
...
+            if settingA:
+                return 'A=True'
+            else:
+                return None
+
+    def create_arguments(filename, file, config_file, settingA: bool):
+        args = ['--lint', '--mode', 'standard']
+        if settingA:
+            return '--use-config', config_file
+        return args
```
Comment (Makman2) [inline docs/Developers/Linter_Bears_Advanced.rst]: wrong indent of  `create_arguments`

File: docs/Developers/Linter_Bears_Advanced.rst [hunk 11]
```diff
@@ -41,6 +40,28 @@ The path of the temporary configuration file can be accessed inside
...
+where to explicitly return ``None`` is when the user wants to use his own
+tool-specific config. If a config file is supplied by user that means we
+don't have to generate a config file. Generating config file consumes
+resources. Avoiding generating config file increases the performance.
+
+::
+
+    @linter(executable='...')
+    class MyBear:
+        def generate_config(filename, file, settingA: bool):
```
Comment (Makman2) [inline docs/Developers/Linter_Bears_Advanced.rst]: maybe call it `user_config` to follow better the explanation above

Comment (Makman2) [reply docs/Developers/Linter_Bears_Advanced.rst]: ah wait `user_config` has to be added together with `settingA` (which should be renamed to `setting_a` btw :smile:)

File: docs/Developers/Linter_Bears_Advanced.rst [hunk 12]
```diff
@@ -41,6 +40,28 @@ The path of the temporary configuration file can be accessed inside
...
+            if settingA:
+                return 'A=True'
+            else:
+                return None
+
+    def create_arguments(filename, file, config_file, settingA: bool):
+        args = ['--lint', '--mode', 'standard']
+        if settingA:
+            return '--use-config', config_file
```
Comment (Makman2) [inline docs/Developers/Linter_Bears_Advanced.rst]: you should incorporate `filename` into the arguments. Also you don't need `--mode` and `standard`, but you always have to append `--use-config` (depending on whether a user config was supplied, put there `user_config` or `config_file`)
-->
```
return filename, '--use-config', user_config if user_config else config_file
```

Comment (Makman2) [reply docs/Developers/Linter_Bears_Advanced.rst]: And then you don't need to provide `settingA` / `setting_a` in the signature altogether :)

**Comment by Makman2** [comment]
unack 1575a86745f4b45307583c89fbfa8563d6d02a64

File: docs/Developers/Linter_Bears_Advanced.rst (hunk 12 above)
Comment (Makman2) [reply docs/Developers/Linter_Bears_Advanced.rst]: only remove `settingA` inside `create_arguments`, you don't need it there :+1: But we need it for `generate_config?` :)

> Also now I'll have to add user_config as input argument. Right?

yes :+1:

File: docs/Developers/Linter_Bears_Advanced.rst [hunk 13]
```diff
@@ -41,6 +40,26 @@ The path of the temporary configuration file can be accessed inside
         def create_arguments(filename, file, config_file):
             return "--use-config", config_file
 
+If you return ``None``, no configuration file is generated. A common case
+where to explicitly return ``None`` is when you want to expose a setting
+for the user to use his own tool-specific config. In case the user specifies
+such a config file, we can avoid generating once again with
```
Comment (Makman2) [inline docs/Developers/Linter_Bears_Advanced.rst]: generating **one**~~once~~

File: docs/Developers/Linter_Bears_Advanced.rst [hunk 14]
```diff
@@ -41,6 +40,26 @@ The path of the temporary configuration file can be accessed inside
...
+``generate_config`` to reduce I/O load.
+
+::
+
+    @linter(executable='...')
+    class MyBear:
+        def generate_config(filename, file, setting_a: bool):
```
Comment (Makman2) [inline docs/Developers/Linter_Bears_Advanced.rst]: we should add `user_config` here too, and do `if user_config`, this is better in case a bear is written with more than one setting (which is likely the case)

Comment (Makman2) [inline docs/Developers/Linter_Bears_Advanced.rst]: You have to annotate `generate_config` with `@staticmethod`, as you don't have `self` in the signature

File: docs/Developers/Linter_Bears_Advanced.rst [hunk 15]
```diff
@@ -41,6 +40,26 @@ The path of the temporary configuration file can be accessed inside
...
+            if setting_a:
+                return 'A=True'
+            else:
+                return None
+
+        def create_arguments(filename, file, user_config, config_file):
```
Comment (Makman2) [inline docs/Developers/Linter_Bears_Advanced.rst]: wrong order of parameters, `config_file` has to be the third parameter. `user_config` should also have a default value of `''` and be annotated with `str`.

File: docs/Developers/Linter_Bears_Advanced.rst [hunk 16]
```diff
@@ -41,6 +40,26 @@ The path of the temporary configuration file can be accessed inside
...
+            return (filename,'--use-config',
+                            user_config if user_config else config_file)
```
Comment (Makman2) [inline docs/Developers/Linter_Bears_Advanced.rst]: Indentation relative to opening brace and missing space after comma:
```
(filename, '--use-config',
 ...)
```

File: docs/Developers/Linter_Bears_Advanced.rst (hunk 14 above)
Comment (Makman2) [reply docs/Developers/Linter_Bears_Advanced.rst]: `setting_a` should also have a default value, like `user_config` when you add it.

Comment (Vedarth) [reply docs/Developers/Linter_Bears_Advanced.rst]: So, should I remove setting_a?

**Comment by Makman2** [comment]
unack e3d388883258bb11e2b56e3f7f376ece846b9bbc

File: docs/Developers/Linter_Bears_Advanced.rst [hunk 17]
```diff
@@ -36,11 +35,33 @@ The path of the temporary configuration file can be accessed inside
             config_file = ("value1 = 1\n"
                            "value2 = 2")
             return config_file
-
+axvtro
```
Comment (sangamcse) [inline docs/Developers/Linter_Bears_Advanced.rst]: ??

Comment (Vedarth) [reply docs/Developers/Linter_Bears_Advanced.rst]: Sorry about that. My bad.

File: docs/Developers/Linter_Bears_Advanced.rst [hunk 18]
```diff
@@ -41,6 +40,28 @@ The path of the temporary configuration file can be accessed inside
...
+for the user to use his own tool-specific config. In case the user specifies
+such a config file, we can avoid generating one again with ``generate_config``
+to reduce I/O load.
+
+::
+
+    @linter(executable='...')
+    class MyBear:
+        @staticmethod
+        def generate_config(filename, file, user_config, setting_a: bool):
```
Comment (sangamcse) [inline docs/Developers/Linter_Bears_Advanced.rst]: I think `user_config` should contain default value here, too.

Comment (sangamcse) [reply docs/Developers/Linter_Bears_Advanced.rst]: Also, `setting_a` should have a default value. Look https://github.com/coala/coala/pull/4899#discussion_r168275071

Comment (Vedarth) [reply docs/Developers/Linter_Bears_Advanced.rst]: Okay. Let me do that.

File: docs/Developers/Linter_Bears_Advanced.rst [hunk 19]
```diff
@@ -41,6 +40,30 @@ The path of the temporary configuration file can be accessed inside
...
+::
+
+    @linter(executable='...')
+    class MyBear:
+        @staticmethod
+        def generate_config(filename, file,
+                            user_config: str = '',
+                            setting_a: str = ''):
+            if user_config:
+                return 'A=True'
```
Comment (Makman2) [inline docs/Developers/Linter_Bears_Advanced.rst]: still you have to return `None` here instead :) Please switch the cases
```
if user_config:
    return None
else:
    return 'A={}'.format(setting_a)
```
Oh and as you see, please incorporate `setting_a` like in my example, so people see why it's there :)

File: docs/Developers/Linter_Bears_Advanced.rst [hunk 20]
```diff
@@ -41,6 +40,30 @@ The path of the temporary configuration file can be accessed inside
...
+to reduce I/O load.
+
+::
+
+    @linter(executable='...')
+    class MyBear:
+        @staticmethod
+        def generate_config(filename, file,
+                            user_config: str = '',
+                            setting_a: str = ''):
```
Comment (Makman2) [inline docs/Developers/Linter_Bears_Advanced.rst]: Optional: Do you want to make it bool? As below there's `A=True`. But the type can be arbitrary.

File: docs/Developers/Linter_Bears_Advanced.rst [hunk 21]
```diff
@@ -41,6 +40,30 @@ The path of the temporary configuration file can be accessed inside
...
+            if user_config:
+                return 'A=True'
+            else:
+                return None
+
+        def create_arguments(filename, file, config_file,
+                            user_config: str = ''):
```
Comment (Makman2) [inline docs/Developers/Linter_Bears_Advanced.rst]: wrong indentation still

File: docs/Developers/Linter_Bears_Advanced.rst [hunk 22]
```diff
@@ -41,6 +40,30 @@ The path of the temporary configuration file can be accessed inside
...
+            return (filename, '--use-config', user_config
+                    if user_config else config_file)
```
Comment (Makman2) [inline docs/Developers/Linter_Bears_Advanced.rst]: see previous comment:
```
return (filename, '--use-config',
        user_config if user_config else config_file)
```

Comment (Makman2) [reply docs/Developers/Linter_Bears_Advanced.rst]: (There's actually a short form, `user_config or config_file`, but let's keep the `if` construct, that's easier to understand)

**Comment by Makman2** [comment]
unack c00331fa1965f393b0034f3fc435b1ca439452df

File: docs/Developers/Linter_Bears_Advanced.rst (hunk 20 above)
Comment (Vedarth) [reply docs/Developers/Linter_Bears_Advanced.rst]: I think making it `bool` will be better.

File: docs/Developers/Linter_Bears_Advanced.rst (hunk 22 above)
Comment (Vedarth) [reply docs/Developers/Linter_Bears_Advanced.rst]: Oh sorry I forgot about that.

**Comment by Vedarth** [comment]
@Makman2 I have made the required changes.

File: docs/Developers/Linter_Bears_Advanced.rst [hunk 23]
```diff
@@ -41,6 +40,30 @@ The path of the temporary configuration file can be accessed inside
...
+to reduce I/O load.
+
+::
+
+    @linter(executable='...')
+    class MyBear:
+        @staticmethod
+        def generate_config(filename, file,
+                            user_config: str = '',
+                            setting_a: bool):
```
Comment (Makman2) [inline docs/Developers/Linter_Bears_Advanced.rst]: default value... :D

File: docs/Developers/Linter_Bears_Advanced.rst [hunk 24]
```diff
@@ -41,6 +40,30 @@ The path of the temporary configuration file can be accessed inside
...
+such a config file, we can avoid generating one again with ``generate_config``
+to reduce I/O load.
+
+::
+
+    @linter(executable='...')
+    class MyBear:
+        @staticmethod
+        def generate_config(filename, file,
+                            user_config: str = '',
```
Comment (Makman2) [inline docs/Developers/Linter_Bears_Advanced.rst]: no spaces around `=` :) (same below)

File: docs/Developers/Linter_Bears_Advanced.rst (hunk 23 above)
Comment (Vedarth) [reply docs/Developers/Linter_Bears_Advanced.rst]: Oops. Sorry about that.

File: docs/Developers/Linter_Bears_Advanced.rst (hunk 24 above)
Comment (Vedarth) [reply docs/Developers/Linter_Bears_Advanced.rst]: Done 👍

**Comment by Makman2** [comment]
ack 8626570b7f6fbb525a2b95205f87e3a8073d2ee9

**Comment by Makman2** [comment]
@gitmate-bot rebase

**Comment by Makman2** [comment]
@gitmate-bot ff
